EQL: Add the ability for EQL to perform CCSes

bpintea opened the pull request:
This introduces the ability for EQL to perform searches on a remote cluster, leveraging ES's CCS capabilities.

The remote cluster needs to be on the same version as the local cluster.

bpintea committed:
Add ability for EQL to perform CCSes

This introduces the ability for EQL to perform searches on a remote
cluster.

bpintea committed:
Fix one plugin test

Fix SqlPluginTests, ClusterSettings class can't and need not be mocked.

bpintea committed:
Restrict remote cluster version to local's

The remote cluster cannot be on a version different than that of local
cluster's.

bpintea committed:
Add remote version verification test

Add Verifier test checking on remote cluster version policy.

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feature/eql-ccs

elasticmachine commented:
Pinging @elastic/es-ql (Team:QL)

costin commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/session/EqlConfiguration.java:
```diff
@@ -39,7 +41,8 @@
                             Map<String, Object> runtimeMappings, List<FieldAndFormat> fetchFields, TimeValue requestTimeout,
-                            IndicesOptions indicesOptions, int fetchSize, String clientId, TaskId taskId, EqlSearchTask task) {
+                            IndicesOptions indicesOptions, int fetchSize, String clientId, TaskId taskId, EqlSearchTask task,
+                            RemoteClusterRegistry remoteClusterRegistry) {
```
See the point I was making before - anything that uses EqlConfiguration now has to deal with RemoteClusterRegistry...

costin commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/util/RemoteClusterRegistry.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.Map;
+
+// Mostly a wrapper for the (final, unmockable) RemoteClusterService.
+public class RemoteClusterRegistry {
```
You could add the verification method here and simply pass its method handle around.

bpintea replied:
Addressed as recommended.

costin commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/EqlTestUtils.java:
```diff
@@ -41,7 +49,16 @@ private EqlTestUtils() {
+    private static final RemoteClusterRegistry remoteClusterRegistry;
+    public static final Verifier TEST_VERIFIER;
+
+    static {
+        remoteClusterRegistry = mock(RemoteClusterRegistry.class);
+        when(remoteClusterRegistry.indicesPerRemoteCluster(any())).thenReturn(Collections.emptyMap());
+        TEST_VERIFIER = testVerifier(new Metrics());
+    }
```
See above - don't mock the infrastructure, mock the logic/result.

costin commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/CancellationTests.java:
```diff
@@ -51,17 +60,32 @@
+    private ThreadPool threadPool;
+    private TransportService transportService;
+
+    @Before
+    public void mockTransportService() {
+        threadPool = new TestThreadPool(getClass().getName());
+        // The TransportService needs to be able to return a valid RemoteClusterServices object down the stream, required by the Verifier.
+        transportService = MockTransportService.createNewService(Settings.EMPTY, Version.CURRENT, threadPool);
+    }
```
Or a protected/package constructor can be declared to simply pass a no-op check for the test suite.

costin commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java:
```diff
@@ -302,7 +302,8 @@ static String getPutMappingIndexOrAlias(PutMappingRequest request, Set<String> a
                 || request instanceof GraphExploreRequest || request instanceof ResolveIndexAction.Request
-                || request instanceof OpenPointInTimeRequest;
+                || request instanceof OpenPointInTimeRequest
+                || request.getClass().getCanonicalName().equals("org.elasticsearch.xpack.eql.action.EqlSearchRequest");
```
This needs to be double checked with the security team.

bpintea replied:
Will do.

bpintea replied:
@albertzaharovits the integration test has been extended to be done with the security enabled, let me know if anything else is missing from your side. Thanks.

albertzaharovits replied:
LGTM to me, thank you @bpintea .

costin reviewed (changes requested):
Looks good overall however there are some improvements around the test setup - the Remote infra classes are too disruptive unnecessary (see the comments for alternatives).

bpintea commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/analysis/Verifier.java:
```diff
@@ -64,9 +67,11 @@
+    private final RemoteClusterRegistry remoteClusterRegistry;
 
-    public Verifier(Metrics metrics) {
+    public Verifier(Metrics metrics, RemoteClusterRegistry remoteClusterRegistry) {
```
My thought was that the `RemoteClusterRegistry` might be later used if we'll want to relax the "same-version-remote-cluster" requirement _or_ do any optimisations like channeling the request through to the remote cluster, if no mixed-cluster indices are given (like `local-index,remote-cluster:remote-index`). So having it in the config might be handy. But will look at alternatives.

bpintea replied:
Refactored inline with Andrei's [recommendation](https://github.com/elastic/elasticsearch/pull/74399#discussion_r657913344) that avoids instantiating the Verifier per session.

bpintea commented on x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/EqlDateNanosSpecTestCase.java:
```diff
@@ -22,7 +22,15 @@
+        super(indexName(), query, name, eventIds);
+    }
+
+    public EqlDateNanosSpecTestCase(String index, String query, String name, long[] eventIds) {
+        super(index, query, name, eventIds);
+    }
+
+    protected static String indexName() {
+        return DATE_NANOS_INDEX;
```
The 2nd constructor is called by the `multi-cluster` suite with the `index` value of remote cluster's name (gradle-defined) plus `DATE_NANOS_INDEX` (i.e. `my_remote_cluster:eql_date_nanos`).

bpintea replied:
Addressed.

costin replied:
Remove indexName() and force the callers to use the static field name instead.

bpintea commented on x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/EqlExtraSpecTestCase.java:
```diff
@@ -25,6 +25,14 @@ public EqlExtraSpecTestCase(String query, String name, long[] eventIds) {
         super(TEST_EXTRA_INDEX, query, name, eventIds);
     }
 
+    public EqlExtraSpecTestCase(String index, String query, String name, long[] eventIds) {
+        super(index, query, name, eventIds);
+    }
+
```
All the test suites shared between the rest and multi-cluster tests are called both with the unqualified and qualified index name, respectively.

bpintea commented on x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/RemoteClusterAwareEqlRestTestCase.java:
```diff
@@ -0,0 +1,109 @@
+    }
+
+    protected static RestClient provisioningClient() {
+        return remoteClusterClient == null ? client() : remoteClusterClient;
```
The `remoteClusterClient` is used to access the "remote-cluster" cluster. `client`/`adminClient` are used to access the local cluster and, in case of the `multi-cluster` tests, perform the CCS searches. That's why the "provisioning" methods. I will add some comments to make it clearer (unless some other design might be better).

bpintea committed:
Address review comments

- add a qualifiedIndex() method to SearchHit;
- rework the way the Verifier checks remote cluster version
  compatibility;
- add more comments.

bpintea committed:
Merge branch 'master' into feature/eql-ccs

Luegg reviewed (approved):
Only a few nitpicks from my side but interesting stuff!

costin commented on x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/EqlSpecTestCase.java:
```diff
@@ -35,6 +35,14 @@ protected String tiebreaker() {
 
     public EqlSpecTestCase(String query, String name, long[] eventIds) {
-        super(TEST_INDEX, query, name, eventIds);
+        super(indexName(), query, name, eventIds);
```
`this(TEST_INDEX, query, name, eventIds)`

bpintea replied:
Addressed.

bpintea committed:
Address review comments

- revert the SearchHit#qualifiedIndex() method;
- more comments;
- renaming some methods and a class;
- relaxing the Collection type for reported version incompatible
  clusters;
- typos and spacing.

astefan commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/index/IndexResolver.java:
```diff
@@ -163,16 +168,24 @@ public boolean equals(Object obj) {
     private final DataTypeRegistry typeRegistry;
 
-    public IndexResolver(Client client, String clusterName, DataTypeRegistry typeRegistry) {
+    private final RemoteClusterResolver remoteClusters;
```
The name of the variable is misleading. Also, why not storing the clusters themselves in the first place, instead of the resolver?

costin replied:
The cluster list is not static since remotes can go offline or be added/removed.

astefan commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/index/IndexResolver.java:
```diff
@@ -163,16 +168,24 @@ public boolean equals(Object obj) {
         return clusterName;
     }
 
+    public Set<String> clusters() {
```
Where is this method used? Leftover?

astefan reviewed (commented):
Looks good overall, but I have some comments and questions.
Also, I am not seeing any more specific tests that make use of the remote clusters.... thinking here of querying just the index from the remote cluster. Or are these "hidden" behind the IT tests in `multi-cluster` project?

bpintea committed:
Address review comments

- switch back to a per PlanExecutor Verifier.
- rip out the dead code (will be added back for SQL) in IndexResolver.
- remove unused prop in gradle script.

bpintea committed:
Merge branch 'master' into feature/eql-ccs

bpintea committed:
Fix method name clash

- secureClientSettings() accidentally incorrectly overridden.

bpintea committed:
Merge branch 'master' into feature/eql-ccs

bpintea commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/payload/EventPayload.java:
```diff
@@ -25,7 +27,7 @@ public EventPayload(SearchResponse response) {
         values = new ArrayList<>(hits.size());
         for (SearchHit hit : hits) {
-            values.add(new Event(hit.getIndex(), hit.getId(), hit.getSourceRef(), hit.getFields()));
+            values.add(new Event(qualifiedIndex(hit), hit.getId(), hit.getSourceRef(), hit.getFields()));
```
Kept as is, to avoid changing server's `SearchHit`.

bpintea commented on x-pack/plugin/eql/qa/multi-cluster/build.gradle:
```diff
@@ -0,0 +1,48 @@
+tasks.register("startRemoteCluster", DefaultTestClustersTask.class) {
+  useCluster testClusters.'remote-cluster'
+  doLast {
+    "Starting remote cluster before integ tests and integTest clustser is started"
```
Fixed.

bpintea commented on x-pack/plugin/eql/qa/multi-cluster/build.gradle:
```diff
@@ -0,0 +1,48 @@
+    nonInputProperties.systemProperty 'tests.rest.cluster.remote.host', "${-> testClusters.'remote-cluster'.getAllHttpSocketURI().get(0)}"
+  }
+}
+tasks.named("check").configure{dependsOn("integTest") } // run these tests as part of the "check" task
```
Thanks, fixed.

bpintea commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/optimizer/TomlFoldTests.java:
```diff
@@ -39,8 +38,7 @@
     private static final EqlFunctionRegistry FUNCTION_REGISTRY = new EqlFunctionRegistry();
-    private static final Verifier VERIFIER = new Verifier(new Metrics());
-    private static final Analyzer ANALYZER = new Analyzer(TEST_CFG, FUNCTION_REGISTRY, VERIFIER);
+    private final Analyzer analyzer = new Analyzer(TEST_CFG, FUNCTION_REGISTRY, TEST_VERIFIER);
```
Change reverted, thanks.

bpintea commented on x-pack/plugin/eql/qa/multi-cluster/src/test/java/org/elasticsearch/xpack/eql/RemoteClusterUtils.java:
```diff
@@ -0,0 +1,25 @@
+package org.elasticsearch.xpack.eql;
+
+public class RemoteClusterUtils {
+    public static final String REMOTE_CLUSTER_NAME = "my_remote_cluster";
```
Added.

bpintea commented on x-pack/plugin/eql/qa/multi-cluster/build.gradle:
```diff
@@ -0,0 +1,48 @@
+tasks.named("integTest").configure {
+  dependsOn 'startRemoteCluster'
+  useCluster testClusters.'remote-cluster'
+  systemProperty 'tests.rest.suite', 'multi_cluster'
```
Removed (c&p issue).

bpintea commented on x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/RemoteClusterAwareEqlRestTestCase.java:
```diff
@@ -0,0 +1,111 @@
+            }
+            String host = crossClusterHost.substring(0, portSeparator);
+            int port = Integer.parseInt(crossClusterHost.substring(portSeparator + 1));
+            HttpHost[] remoteHttpHosts = new HttpHost[]{new HttpHost(host, port)};
```
There seems to be some inconsistency in ES around this particular piece of style. I've applied your suggestion, though.

bpintea commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/PlanExecutor.java:
```diff
@@ -54,13 +53,13 @@ public PlanExecutor(Client client, IndexResolver indexResolver, NamedWriteableRe
     private EqlSession newSession(EqlConfiguration cfg) {
-        return new EqlSession(client, cfg, indexResolver, preAnalyzer, postAnalyzer, functionRegistry, verifier, optimizer, planner, this);
+        return new EqlSession(client, cfg, indexResolver, preAnalyzer, postAnalyzer, functionRegistry,
+            new Verifier(metrics, cfg.versionIncompatibleClusters()), optimizer, planner, this);
```
Indeed, good point, thanks! Updated accordingly.

bpintea commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/util/RemoteClusterRegistry.java:
```diff
@@ -0,0 +1,52 @@
+
+    private Version remoteVersion(String clusterAlias) {
+        return remoteClusterService.getConnection(clusterAlias).getVersion();
+    }
```
Applied.

bpintea commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/index/IndexResolver.java:
```diff
@@ -163,16 +168,24 @@ public boolean equals(Object obj) {
         this.client = client;
         this.clusterName = clusterName;
         this.typeRegistry = typeRegistry;
+        this.remoteClusters = new RemoteClusterResolver(settings, clusterSettings);
```
This, and the code referenced by the below two comments has been now removed.

bpintea commented on x-pack/plugin/eql/qa/multi-cluster-with-security/src/test/java/org/elasticsearch/xpack/eql/RemoteClusterTestUtils.java:
```diff
@@ -0,0 +1,25 @@
+    public static final String REMOTE_CLUSTER_NAME = "my_remote_cluster"; // gradle defined
+
+    public static String remoteClusterIndex(String indexName) {
+        return REMOTE_CLUSTER_NAME + ":" + indexName;
```
`RemoteClusterAware#buildRemoteIndexName()` not available here.

bpintea commented:
> Or are these "hidden" behind the IT tests in multi-cluster project?

Indeed, the `multi-cluster-with-security` is running all the rest tests, but in a CCS fashion.

costin commented on x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/EqlExtraSpecTestCase.java:
```diff
@@ -21,8 +21,18 @@
+        super(index, query, name, eventIds);
+    }
+
+    protected static String indexName() {
```
See above.

costin commented on x-pack/plugin/eql/qa/common/src/main/java/org/elasticsearch/test/eql/EqlSpecTestCase.java:
```diff
@@ -34,7 +34,17 @@ protected String tiebreaker() {
+        super(index, query, name, eventIds);
+    }
+
+    protected static String indexName() {
```
Same thing - this method is not needed.

bpintea replied:
Addressed now.

costin commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/analysis/Verifier.java:
```diff
@@ -70,11 +74,11 @@ public Verifier(Metrics metrics) {
 
     public Map<Node<?>, String> verifyFailures(LogicalPlan plan) {
-        Collection<Failure> failures = verify(plan);
+        Collection<Failure> failures = verify(plan, x -> Collections.emptySet());
```
Why use an emptySet when verifying failures vs verify?

bpintea replied:
The function doesn't seem to be used anywhere, I assumed it's a debugging left-over. I've removed it now as dead code.

costin commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/util/RemoteClusterRegistry.java:
```diff
@@ -0,0 +1,52 @@
+        return indicesMap;
+    }
+
+    private Version remoteVersion(String clusterAlias) {
```
Nit - since it's a one liner, no need to define a separate method since it's only used once.

bpintea replied:
Line inlined now.

costin commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/analysis/CancellationTests.java:
```diff
@@ -227,4 +250,8 @@ private ClusterService mockClusterService(String nodeId) {
     }
+
+    private static IndexResolver indexResolver(Client client) {
+        return new IndexResolver(client, randomAlphaOfLength(10), DefaultDataTypeRegistry.INSTANCE);
```
👍

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/index/IndexResolver.java:
```diff
@@ -795,4 +800,24 @@ private GetAliasesRequest createGetAliasesRequest(FieldCapabilitiesResponse resp
         return emptyMap();
     }
+
+    private class RemoteClusterResolver extends RemoteClusterAware {
```
Can be made static as well.

bpintea replied:
I've removed the class entirely, as the calling code has been removed. (It'll be re-added as static class for the SQL CCS.)

costin commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/analysis/CancellationTests.java:
```diff
@@ -231,4 +231,8 @@ private ClusterService mockClusterService(String nodeId) {
         return mockClusterService;
     }
+
+    private static IndexResolver indexResolver(Client client) {
```
Since it's another test where this code is used, how about moving it in EqlTestUtils.

bpintea replied:
I think there's a similar call in SQL's cancelation tests only (hope I didn't miss other place), but not sure how this code could be shared between them.

costin reviewed (approved):
Glad to see the Verifier being kept as a singleton.
Left some minor comments, LGTM.

bpintea committed:
Address review comments

- replace function with constant in test code.
- remove unused function.

bpintea commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into feature/eql-ccs

bpintea committed:
Address review comments

- remove unused class;
- replace single use of one-liner function with actual function line.

bpintea committed:
Remove redundant cluster configs

Remove redundant (and overriden) security config lines.

bpintea committed:
Merge branch 'master' into feature/eql-ccs

astefan reviewed (approved):
LGTM
